### ISSUE TITLE
Build and test coverage: prebuilt images on docker hub

### ISSUE THREAD

**Issue by kdave** [kdave/btrfs-progs#63]

**Comment by Lakshmipathi** [comment]
Here is the initial files for Fedora,CentOS,Ubuntu and OpenSUSE. https://github.com/Lakshmipathi/btrfs-docker-images

While trying to run 'make test' I got this error: _mount: /btrfs-progs/tests/fsck-tests/012-leaf-corruption/test.img: failed to setup loop device: No such file or directory_
May be its related to bug: _https://github.com/moby/moby/issues/27886_ 

Will check this issue before pushing the images to Docker hub. Let me know if you have preferred distro or/and architectures. I'm not quite familiar with travis, will look into it later.

**Comment by kdave** [comment]
You can take the docker/travis integration from here, https://github.com/kdave/btrfs-progs/pull/17/commits/6ec1ce8eacb6d001e24c3d419cb235464e0822d1 .

Initially I think adding plain build test for the distros, the four you have now are ok. The tests can take long and I'd limit it to running them on push to devel (as is now) on the main architectures, but release tests can run on all distro:arch pairs.

Architectures:
* x86_64 is already tested, but would to be good for build test anyway
* arm 32bit and arm 64bit
* maybe: ppc64, mips

**Comment by Lakshmipathi** [comment]
okay. I got the loop device working, when the container launched with --privileged  and --device options. I can run 'make test' now without loop-back device errors.

**Comment by kdave** [comment]
Thanks for the update. I was wondering if we actually need all the distros covered, as there are very few things where it would matter and I believe most of them would be solved on the packaging level. My biggest interest is in testing the non-intel architectures and for that purpose we can simply pick any distro that provides the architecture. Otherwise, say we have 4 distros and 4 arches, that's going to be 16 images to manage and pull into travis. Not only this would take a long time to finish but also does not significanly enhance the tests.

**Comment by Lakshmipathi** [comment]
Yes, you are correct, covering all distro may be not relevant.  Will first work for ARM build for one of those distro (probably CentOS) and then gradually move to others.

**Comment by Lakshmipathi** [comment]
Here is the ARMv8 Docker image. Dockerfiles can be found at git repo.

Seems like ARM image can be built from Intel device, but its time consuming and found it quite difficult. 
so used https://www.scaleway.com/instantcloud to build/test/push images. I ran only few tests(fsck,convert) in those 15 minutes intervals, If you find issues, let me know. 


```
docker pull laks/btrfs_armv8
docker run -it --privileged=true --device=/dev/loop0:/dev/loop0 --device=/dev/loop1:/dev/loop1  --device=/dev/loop2:/dev/loop2  --device=/dev/loop3:/dev/loop3 --device=/dev/loop4:/dev/loop4  laks/btrfs_armv8
```

Will continue to look into other ARCH and Travis integration.

**Comment by kdave** [comment]
I've tried to copy the integration from the original commit but it does not work, https://github.com/kdave/btrfs-progs/tree/test-travis-docker , errors in https://travis-ci.org/kdave/btrfs-progs/jobs/284302723 .

**Comment by Lakshmipathi** [comment]
okay thanks. I'm exploring travis-ci and checking existing btrfs-builds(https://travis-ci.org/Lakshmipathi/btrfs-progs).  Will have an update on this error soon.

**Comment by Lakshmipathi** [comment]
For travis/docker/arm to work we need to setup QEMU, some how the existing CentOS image failed to work. So switched to ARM-V6 and its working now https://travis-ci.org/Lakshmipathi/btrfs-progs/jobs/284627908

Here's the Docker setup for Travis: https://github.com/Lakshmipathi/btrfs-progs/blob/test-travis-docker/.travis.yml

Will check ARM-v8. We need to include test instead of  'btrfs --version'.

**Comment by Lakshmipathi** [comment]
Now Added ARMv8:
https://travis-ci.org/Lakshmipathi/btrfs-progs/jobs/285617135 
https://github.com/Lakshmipathi/btrfs-progs/blob/test-travis-docker/.travis.yml
https://github.com/Lakshmipathi/btrfs-docker-images/tree/master/arm

**Comment by kdave** [comment]
Great, thanks. The commandlines get pretty long so we'll need to add some helper scritps to keep .travis.yml sane, but otherwise the core of the work is done.
The frequent pushes to devel can do only build checks and the release-test branch can do the long running tests. For sake of documentation and reproducibility, all the dockerfiles should be stored in btrfs-progs, but for the time being we can use your docker hub repository. As the file count will grow, I suggest to add a new directory.

**Comment by kdave** [comment]
The integration can be now found in branch travis-docker-test, but it still fails during the buid https://travis-ci.org/kdave/btrfs-progs/jobs/286106330 .

**Comment by Lakshmipathi** [comment]
Okay sure, we can move the dockerfiles into btrfs-progs once travis/docker build starts working fine. 

Thanks for the quick update and script :)  I looked into the failure, seems like docker image missing zstd (and also reiserfs-progs) . I think, the stuff we do inside "before_install" performed in host system so they are missing from docker image. We see configure saying  "btrfs-restore zstd: no" for docker builds.  Will add these missing packages into docker image and update you.

**Comment by kdave** [comment]
Right now zstd and reiserfs are optional so the build should work without them, the error looks like some common include header cannot be found but I don't see why.

**Comment by Lakshmipathi** [comment]
okay, I was looking at test error. Yes you are right built itself failing with message like 
`make: *** No rule to make target '/home/travis/build/kdave/btrfs-progs/kernel-lib/list.h', needed by 'btrfs.o'.  Stop.`

Not sure whats going on there, will check.

**Comment by kdave** [comment]
This comes from the phase where make dependencies are generated, ie. the source.o.d files, and the docker run is in a different path `/btrfs-progs` while i report the default build path `/home/travis/build/kdave/btrfs-progs/` . Maybe there are some lefovers in the docker image, but there are only build dependencies installed. Still confused.

**Comment by Lakshmipathi** [comment]
I copied your files and checked for arm6. Build is passed.  Can you try again now with this config file: 
https://github.com/Lakshmipathi/btrfs-progs/blob/test-travis-docker/.travis.yml
https://github.com/Lakshmipathi/btrfs-progs/blob/test-travis-docker/travis/docker-arch-run (added 'shift to ignore first arg)
travis Log:
https://travis-ci.org/Lakshmipathi/btrfs-progs/jobs/286149789

Will trigger build for armv8 and update soon.

**Comment by Lakshmipathi** [comment]
Here's build log for armv8 & v6: 
https://travis-ci.org/Lakshmipathi/btrfs-progs/jobs/286155590
Build is passing on this runs. Will  fork & check your repo tomorrow.

**Comment by kdave** [comment]
I've picked your changes, the shift was my oversight but still the failure in my travis repo is one step before (`make V=1`). I think the only difference is the build cache, so I'll try to disable it and see.

**Comment by Lakshmipathi** [comment]
okay sure.

**Comment by kdave** [comment]
Aha, one more thing: in your travis script, there's no `./autogen && configure && make` in the `before_install` section.

**Comment by Lakshmipathi** [comment]
Yes, may be thats the cache creating some trouble. On my travis script, I removed those cmds, since they prepares the host machine for non-docker build/test. I think without those 'before_install' part for host machine, this issue may go away.

**Comment by kdave** [comment]
Ok, now it works. The autogen && configure is in the before_build phase so it does not pollute the logs. I also don't want to run the arm builds as first steps (ie. move the configure phase after them), but do something like that:
* autogen && configure && make
* make test-cli, test-mkfs, ... all the quicker ones
* run arm builds
* make test-convert (optional, slow)

Temporarily we can let the arm builds run first, but ideally I'd like to be able to reorder the tests based on the overall runtime.

**Comment by Lakshmipathi** [comment]
okay great! That was a tricky issue :+1:  Yes, we can reorder tests and keep arm builds lower priority than x86_64 

If I'm not wrong, we still have one issue, the test (TEST_LOG=dump test-cli) didn't get triggered after arm build is done.

**Comment by Lakshmipathi** [comment]
test-cli is getting triggered if add quotes on **.travis.yml** file like:

> travis/docker-arch-run arm6 "make TEST_LOG=dump test-cli"

instead of 

> travis/docker-arch-run arm6 make TEST_LOG=dump test-cli

**Comment by kdave** [comment]
Thanks, though I don't understand why it does not work without the quotes. I think now we have everything, but quite slow and inefficient. The docker builds are not cached and take a lot of time, so we could possibly insert `ccache` to the system images and then build via scripts that utilize ccache and then save the caches at the end of run.
Next thing I've just noticed are the tests. As the running kernel is x86_64 and binaries are arm, the endianity does not match and ioctls will fail because we assume that endianity of userspace matches kernel. While this is possible to fix, it is pretty uncommon setup. Kernel and userspace should match if we want to run the full tests, so we'll have just the build tests for now.

**Comment by kdave** [comment]
So not everything is done, https://travis-ci.org/kdave/btrfs-progs/jobs/286526374, the docker builds run, the tests expectedly fail and there's the autogen && configure phase that fails. Possibly due to some clashing files.
Seems still too much hassle to get it working, so I'll probably push everything what we've managed to do so far but will not enable the docker part in the config. That way it should be easier to continue debugging.

**Comment by Lakshmipathi** [comment]
okay.  I tried build process failure with "./autogen.sh && ./configure --disable-documentation && make"
Seems like adding sudo made some progress it failed on make  "sudo ./autogen.sh && sudo ./configure --disable-documentation &&  make"   https://travis-ci.org/Lakshmipathi/btrfs-progs/jobs/286564492 Will continue debugging. 

On test coverage:  yes, we can include it later, may be even identify tests we want to run in this 
travis/docker/arm environment. If we can add build tests that should be fine.

**Comment by Lakshmipathi** [comment]
Reproduced that caching issue locally.  

```
$sudo su 
#./autogen.sh && ./configure 
```
Then now leave root shell & performing `./autogen && ./configure `as normal user creates the issue. 
`"make clean"` fails to remove **"autom4te.cache"** dir which is owned by root.
Apparently `sudo make clean` didn't touch **"autom4te.cache"**  too. We need to call `sudo make clean-all`

Below travis link has successful docker and regular builds. 
https://travis-ci.org/Lakshmipathi/btrfs-progs/jobs/286598719

**Comment by kdave** [comment]
Ritght. Would it be possible to run docker in separate directories? Ie with the use of `$SRC_DIR`. The git source code would need to be duplicated, but this could avoid problems with switching from root to normal user.

**Comment by Lakshmipathi** [comment]
okay let me check. I think we need to copy files from $TRAVIS_BUILD_DIR to new location with-in host and then export that directory to container.

**Comment by Lakshmipathi** [comment]
Added REPO_DIR under 'before_install' in travis.yml file.  Then set REPO_DIR  on docker-arch-run used it with docker-run  --volume "$REPO_DIR:$SRC_DIR" cmd. Now docker build doesn't conflict with others.

https://github.com/Lakshmipathi/btrfs-progs/blob/test-travis-docker/.travis.yml
https://github.com/Lakshmipathi/btrfs-progs/blob/test-travis-docker/travis/docker-arch-run

**Comment by kdave** [comment]
I got the travis/docker build working now. We probably can't do the functional tests on other architectures than x86_64, the're are problems with endianity of userspace and kernel. We could still add the distro x arch build tests but this has never been reported as problematic so I don't see a strong need to have them.

**Comment by Lakshmipathi** [comment]
okay nice, then I think this issue can be closed? May be re-open later depending on specific needs?
